csharp/src/Google.Protobuf/Reflection/CustomOptions.cs

```diff
@@ -65,10 +65,15 @@ public sealed class CustomOptions
     {
         private static readonly object[] EmptyParameters = new object[0];
         private readonly IDictionary<int, IExtensionValue> values;
+        // since we made the descriptor protos public, we can't really reintroduce the CustomOptions
```

Comment:
qq: before unknown field support was added, how did the custom options work?
Also how did they work after unknownField support was added but before support for extensions was added?

Understanding these seems important for figuring out the right approach to this.

Reply:
Before CustomOptions worked by acting as a sort of custom UnknownFieldSet with exposed methods to read values. Once unknown field support was added they kept their custom behavior since unknown fields weren't exposed in the message API.